secp256k1-sys/src/lib.rs

```diff
@@ -673,6 +672,80 @@ mod fuzz_dummy {
         fn rustsecp256k1_v0_4_1_context_preallocated_clone(cx: *const Context, prealloc: *mut c_void) -> *mut Context;
     }
 
+    mod once {
+        use core::sync::atomic::{AtomicUsize, Ordering};
+
+        const NONE: usize = 0;
+        const WORKING: usize = 1;
+        const DONE: usize = 2;
+
+        pub(crate) struct Once(AtomicUsize);
+
+        impl Once {
+            pub(crate) const INIT: Once = Once(AtomicUsize::new(NONE));
+
+            pub(crate) fn run(&self, f: impl FnOnce()) {
+                // Acquire ordering because if it's DONE the following reads must go after this load.
+                let mut have_ctx = self.0.load(Ordering::Acquire);
+                if have_ctx == NONE {
+                    // Ordering: on success we're only signalling to other thread that work is in progress
```

Comment:
IIUC this can cause the function to be called twice, which isn't acceptable.

Reply:
Ordering is about relative order of atomic and non-atomic operations. There's no way `compare_exchange` could succeed twice with code like this. Even if the ordering was relaxed.

Reply:
Why not? My understanding is that a `Relaxed` write is not necessarily guaranteed to come before an `Acquire` read - ie if two threads hit this at the same time, the first can `Acquire` the NONE, but with a `Relaxed` write there is no happens-before relationship with the second thread that `Acquire`s the value. It's been a a few years since I stared at the atomic ordering things, though.

Reply:
compare_exchange is atomic operation - either the value was `NONE` and it succeeded and thus there's no longer `NONE` for other thread to see and succeed again or the operation fails because there's something else than `NONE`. Orderings don't matter for atomic operations. Only for combination of atomic and non-atomic.

Reply:
This is definitely not true, see https://en.cppreference.com/w/c/atomic/memory_order#Relaxed_ordering

Reply:
> Orderings don't matter for atomic operations.

Note that this *Is* true on x86, it just isn't true in general or on other platforms. On x86 everything but SeqCst is all the same as non-atomic anyway, though, so short of using uint128 atomic doesn't matter on x86 anyway :).

Reply:
> They only guarantee atomicity and modification order consistency.

This explicitly points out atomicity **is** guaranteed and `compare_exchange` is definitely atomic. It's literally a CPU instruction that performs comparing and swapping in a way that prevents race conditions with other cores. Without this locks would be impossible to implement.

> Note that this Is true on x86

Ordering doesn't affect just CPU instructions but also compiler optimizations.

Given this code:

```rust
let x = X.load(Ordering::Relaxed);
let y = Y;
```

The optimizer can change it as if it was written like this:

```rust
let y = Y;
let x = X.load(Ordering::Relaxed);
```

However if the only thing you need to transfer is the information fully contained in the atomic (e.g. a counter, flag) then ordering doesn't matter for correctness but using anything else than `Relaxed` is waste of resources.

Reply:
> This explicitly points out atomicity is guaranteed and compare_exchange is definitely atomic.

Ah, man it *has* been a while since I've re-read the C atomic orderings, yea, no, you're right here.